File: dotnet/gui/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace PubSubSQLGUI
{
    public partial class MainForm : Form
    {
        private string DEFAULT_ADDRESS = "localhost:7777";
        private int PUBSUB_TIMEOUT = 5; // in milliseconds
        private int FLASH_TIMER_INTERVAL = 150;
        private long FLASH_TIMEOUT = 300 * 10000; // ticks converted to milliseconds
        private PubSubSQL.Client client = PubSubSQL.Factory.NewClient();
        private bool cancelExecuteFlag = false;
        private string connectedAddress = string.Empty;
        private ListViewDataset dataset = new ListViewDataset();
        private Timer flashTimer = new Timer();
        private Simulator simulator = new Simulator();

        public MainForm()
        {
            InitializeComponent();
            // set up controls and events
            setControls(newButton, newMenu, new_);
            exitMenu.Click += exit;
            connectLocalButton.ToolTipText = "Connect to " + DEFAULT_ADDRESS;
            setControls(connectLocalButton, connectLocalMenu, connectLocal);
            connectLocalMenu.Text = connectLocalMenu.ToolTipText;
            setControls(connectButton, connectMenu, connect);
            setControls(disconnectButton, disconnectMenu, disconnect);
            setControls(executeButton, executeMenu, execute);
            setControls(cancelButton, cancelMenu, cancelExecute);
            simulateMenu.Click += simulate;
            aboutMenu.Click += about;
            updateConnectedAddress(string.Empty);
            resultsTabContainer.SelectedTab = statusTab;
            enableDisableControls();

            flashTimer.Interval = FLASH_TIMER_INTERVAL;
            flashTimer.Enabled = true;
            flashTimer.Tick += tick;
        }

        private void enableDisableControls()
        {
            bool connected = client.Connected();
            connectLocalButton.Enabled = !connected;
            connectLocalMenu.Enabled = !connected;
            connectButton.Enabled = !connected;
            connectMenu.Enabled = !connected;
            disconnectButton.Enabled = connected;
            disconnectMenu.Enabled = connected;
            executeButton.Enabled = connected;
            executeMenu.Enabled = connected;
            simulateMenu.Enabled = executeMenu.Enabled;
            cancelButton.Enabled = false;
            cancelMenu.Enabled = false;
        }

        private void setControls(ToolStripButton button, ToolStripMenuItem menu, EventHandler click)
        {
            menu.ToolTipText = button.ToolTipText;
            menu.Click += click;
            button.Click += click;
        }

        // gui events

        private void new_(object sender, EventArgs e)
        {
            System.Diagnostics.Process.Start(Application.ExecutablePath, connectedAddress); 
        }

        private void exit(object sender, EventArgs e)
        {
            this.Close();
        }

        private void connectLocal(object sender, EventArgs e)
        {
            connect(DEFAULT_ADDRESS); 
        }

        private void connect(object sender, EventArgs e)
        {
            ConnectForm dlg = new ConnectForm();
            if (dlg.ShowDialog(this) == System.Windows.Forms.DialogResult.OK)
            {
                connect(dlg.Address);
            }
        }

        private void connect(string address)
        {
            clearResults();
            if (client.Connect(address))
            {
                updateConnectedAddress(address);
            }
            setStatus();
            enableDisableControls();
        }

        private void disconnect(object sender, EventArgs e)
        {
            simulator.Stop();
            cancelExecuteFlag = true;
            updateConnectedAddress(string.Empty);
            client.Disconnect();
            enableDisableControls();
            clearResults();
        }

        private void execute(object sender, EventArgs e)
        {
            try
            {
                executing();
                string command = queryText.Text.Trim();
                if (string.IsNullOrEmpty(command)) return;
                client.Execute(command);
                processResponse();
            }
            finally
            {
                doneExecuting();
                // we were stoped in the middle
                if (cancelExecuteFlag)
                {
                    // lets make it simple and just reconnect
                    // done in order to ignore possible pubsub event or not fully read result set
                    if (!string.IsNullOrEmpty(connectedAddress))
                    {
                        connect(connectedAddress);
                        clearResults();
                    }
                }
            }
        }

        private void cancelExecute(object sender, EventArgs e)
        {
            simulator.Stop();
            cancelExecuteFlag = true;
        }

        private void simulate(object sender, EventArgs e)
        {
            SimulatorForm dlg = new SimulatorForm();
            if (dlg.ShowDialog(this) == DialogResult.OK)
            {
                simulator.Stop();
                simulator.Address = connectedAddress;
                simulator.Columns = dlg.Columns;
                simulator.Rows = dlg.Rows;
                simulator.TableName = dlg.TableName;
                simulator.Start();
                queryText.Text = "subscribe * from " + dlg.TableName;
                execute(sender, e);
            }
        }

        private void about(object sender, EventArgs e)
        {
            AboutForm dlg = new AboutForm();
            dlg.ShowDialog(this);
        }

        long flashTicks = DateTime.Now.Ticks;
        private void tick(object sender, EventArgs e)
        {
            if (dataset.ResetDirty())
            {
                setStatus();
                setJSON();
                listView.VirtualListSize = dataset.RowCount;
                resultsTabContainer.SelectedTab = resultsTab;
                listView.BeginUpdate();
                listView.EndUpdate();
                flashTicks = DateTime.Now.Ticks;
            }
            else if ((DateTime.Now.Ticks - flashTicks) < (FLASH_TIMEOUT * 2) )
            {
                // make sure that we clear the background color
                listView.VirtualListSize = dataset.RowCount;
                listView.BeginUpdate();
                listView.EndUpdate();
            }
        }

        // helper functions

        private void clearResults()
        {
            listView.Columns.Clear();
            listView.VirtualListSize = 0;
            statusText.Text = "";
            jsonText.Text = "";
        }

        private void setStatus()
        {
            if (client.Ok())
            {
                statusText.ForeColor = Color.Black;
                statusText.Text = "ok";
                return;
            }
            statusText.ForeColor = Color.Red;
            statusText.Text = "error\r\n" + client.Error();
            enableDisableControls();
            return;
        }

        private void setJSON()
        {
            jsonText.Text = client.JSON();
        }
        
        private void updateConnectedAddress(string address)
        {
            Text = "PubSubSQL Interactive Query " + address;
            connectedAddress = address;
        }

        private void executing()
        {        
            clearResults();
            cancelExecuteFlag = false;
            queryText.Enabled = false;
            executeButton.Enabled = false;
            executeMenu.Enabled = false;
            simulateMenu.Enabled = false;
            cancelButton.Enabled = true;
            cancelMenu.Enabled = true;
        }

        private void doneExecuting()
        {
            queryText.Enabled = true;
            enableDisableControls();
        }

        private void processResponse()
        {
            dataset.Clear();
            // determine if we just subscribed  
            if (client.PubSubId() != string.Empty && client.Action() == "subscribe")
            {
                setStatus();
                setJSON();
                // enter event loop
                waitForPubSubEvent();
                return;
            }
            // check if it is result set
            if (client.RowCount() > 0 && client.ColumnCount() > 0)
            {
                updateDataset(); 
                resultsTabContainer.SelectedTab = resultsTab;
            }
            //            
            if (client.Failed())resultsTabContainer.SelectedTab = statusTab;
            setStatus();
            setJSON();
        }

        private void syncColumns()
        {
            for (int i = listView.Columns.Count; i < dataset.ColumnCount; i++)
            {
                string col = dataset.Column(i); 
                listView.Columns.Add(col);
                listView.Columns[i].Width = 100;
            }
        }

        private void waitForPubSubEvent()
        {
            while (!cancelExecuteFlag)
            {
                bool timedout = !client.WaitForPubSub(PUBSUB_TIMEOUT);
                if (client.Failed()) break;
                if (!timedout) updateDataset();
                Application.DoEvents();
            }
            if (client.Failed()) setStatus();
        }

        private void updateDataset()
        {
            if (!(client.RowCount() > 0 && client.ColumnCount() > 0)) return;
            // inside dataset
            dataset.SyncColumns(client);
            syncColumns();
            dataset.AddRowsCapacity(client.RowCount());
            while (client.NextRow() && !cancelExecuteFlag)
            {
                dataset.ProcessRow(client);
            }
        }

        private void listView_RetrieveVirtualItem(object sender, RetrieveVirtualItemEventArgs e)
        {
            List<Cell> row = dataset.GetRow(e.ItemIndex);
            ListViewItem item = new ListViewItem();
            item.UseItemStyleForSubItems = false;
            for (int i = 0; i < listView.Columns.Count; i++)
            {
                string str = string.Empty;
                Color cellBackColor = Color.White;
                if (i < row.Count)
                {
                    Cell cell = row[i];
                    str = cell.Value;
                    if ((DateTime.Now.Ticks - cell.LastUpdated) < FLASH_TIMEOUT)
                    {
                        cellBackColor = Color.HotPink;
                    }
                }
                if (i == 0)
                {
                    item.Text = str;
                }
                else
                {
                    item.SubItems.Add(str).BackColor = cellBackColor;
                }
            }
            e.Item = item;
        }

        private void MainForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            cancelExecuteFlag = true;
            simulator.Stop();
            client.Disconnect();
        }

        private void MainForm_Load(object sender, EventArgs e)
        {
            string[] args = Environment.GetCommandLineArgs();
            if (args.Length > 1)
            {
                connect(args[1]);
            }
        }

    }
}

File: java/gui/MainForm.java
/* Copyright (C) 2014 CompleteDB LLC.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with PubSubSQL.  If not, see <http://www.gnu.org/licenses/>.
 */

import java.awt.*;
import java.awt.event.*;
import javax.swing.*;

public class MainForm extends JFrame implements ActionListener {

	private JMenuItem connectLocalMenu;
	private JButton connectLocalButton;
	private JMenuItem connectMenu;
	private JButton connectButton;
	private JMenuItem disconnectMenu;
	private JButton disconnectButton;
	private JMenuItem executeMenu;
	private JButton executeButton;
	private JMenuItem cancelMenu;
	private JButton cancelButton;
	private JMenuItem simulateMenu;
		
	private JTextArea queryText;
	private JTabbedPane resultsTabContainer;
	private JTextArea statusText;
	private JTextArea jsonText;
	private String DEFAULT_ADDRESS = "localhost:7777";
	private pubsubsql.Client client = pubsubsql.Factory.NewClient();
	private String connectedAddress = "";
	private boolean cancelExecuteFlag = false;
	private TableDataset dataset = new TableDataset();
	private int FLASH_TIMER_INTERVAL = 150;	
	private int PUBSUB_TIMEOUT = 5;
	private TableView tableView = new TableView(FLASH_TIMER_INTERVAL * 2000000, dataset); 
	private Timer timer;	
	private Simulator simulator = new Simulator();

	private AboutForm aboutForm;
	private ConnectForm connectForm;

	public MainForm() {
		Toolkit toolkit = Toolkit.getDefaultToolkit();
        Dimension screen = toolkit.getScreenSize();
		setupMenuAndToolBar();		
		// query text
		queryText = new JTextArea();
		queryText.setPreferredSize(new Dimension(screen.width / 2, 100));
		// tabs
		resultsTabContainer = new JTabbedPane();
		resultsTabContainer.addTab("Results", tableView);
		statusText = new JTextArea();		
		resultsTabContainer.addTab("Status", statusText);
		jsonText = new JTextArea();
		resultsTabContainer.addTab("JSON Response", jsonText);
		// splitter
		JSplitPane splitPane = new JSplitPane(JSplitPane.VERTICAL_SPLIT, queryText, resultsTabContainer); 
		this.add(splitPane, BorderLayout.CENTER);	
		// position
        setSize(screen.width / 2, screen.height / 2);
        setLocation(screen.width / 4, screen.height / 4);
		//
        updateConnectedAddress("");
		enableDisableControls();
		//
		timer = new Timer(FLASH_TIMER_INTERVAL, this); 
		timer.start();
	}

	void setupMenuAndToolBar() {
		JMenuBar menuBar = new JMenuBar();
		setJMenuBar(menuBar);	
		JToolBar toolBar = new JToolBar();
		add(toolBar, BorderLayout.NORTH);
		// File
		JMenu fileMenu = new JMenu("File");
			// New
			JMenuItem newMenu = new JMenuItem(new_);	
			new_.putValue(Action.SHORT_DESCRIPTION, "New PubSubSQL Interactive Query");
			fileMenu.add(newMenu);
			fileMenu.addSeparator();
			toolBar.add(new_);
			toolBar.addSeparator();
			// Exit	
			JMenuItem exitMenu = new JMenuItem(exit);
			defaultTooltips(exit);
			fileMenu.add(exitMenu);
		menuBar.add(fileMenu);
		// Connection
		JMenu connectionMenu = new JMenu("Connection");
			// Connect local
			connectLocalMenu = new JMenuItem(connectLocal);
			defaultTooltips(connectLocal);
			connectionMenu.add(connectLocalMenu);
			connectLocalButton = toolBar.add(connectLocal);
			// Connect
			connectMenu = new JMenuItem(connect);
			connect.putValue(Action.SHORT_DESCRIPTION, "Connect to remote server");
			connectionMenu.add(connectMenu);
			connectButton = toolBar.add(connect);
			// Disconnect
			disconnectMenu = new JMenuItem(disconnect);
			defaultTooltips(disconnect);
			connectionMenu.add(disconnectMenu);
			disconnectButton = toolBar.add(disconnect);
			toolBar.addSeparator();
		menuBar.add(connectionMenu);	
		// Query
		JMenu queryMenu = new JMenu("Query");
			// Execute 
			executeMenu = new JMenuItem(execute);
			defaultTooltips(execute);
			queryMenu.add(executeMenu);
			executeButton = toolBar.add(execute);
			// Cancel Executing Query 
			cancelMenu = new JMenuItem(cancelExecute);
			defaultTooltips(cancelExecute);
			queryMenu.add(cancelMenu);
			cancelButton = toolBar.add(cancelExecute);
			// Simulate 
			simulateMenu = new JMenuItem(simulate);
			defaultTooltips(simulate);
			queryMenu.add(simulateMenu);
		menuBar.add(queryMenu);	
		// Help
		JMenu helpMenu = new JMenu("Help");
			// About 
			JMenuItem aboutMenu = new JMenuItem(about);
			defaultTooltips(about);
			helpMenu.add(aboutMenu);
		menuBar.add(helpMenu);	
	}

	// timer event
	long flashTicks = System.nanoTime();
	public void actionPerformed(ActionEvent e) {
		if (dataset.ResetDirtyData()) {
			setStatus();
			setJSON();
			tableView.Update();
			flashTicks = System.nanoTime();
		} else if (System.nanoTime() - flashTicks < tableView.FLASH_TIMEOUT * 2) {
			tableView.Update();
		}
	}

	// events
	Action new_ = new AbstractAction("New", createImageIcon("images/New.png")) {
		public void actionPerformed(ActionEvent event) {
			
		}
	};
	
	Action exit = new AbstractAction("Exit") {
		public void actionPerformed(ActionEvent event) {
			System.exit(0);
		}
	};

	Action connectLocal = new AbstractAction("Connect to " + DEFAULT_ADDRESS, createImageIcon("images/ConnectLocal.png")) {
		public void actionPerformed(ActionEvent event) {
			connect(DEFAULT_ADDRESS);
		}
	};

	Action connect = new AbstractAction("Connect...", createImageIcon("images/Connect.png")) {
		public void actionPerformed(ActionEvent event) {
			if (connectForm == null) connectForm = new ConnectForm(MainForm.this);	
			connectForm.setLocationRelativeTo(MainForm.this);
			connectForm.setHost("localhost");	
			connectForm.setPort(7777);
			connectForm.setVisible(true);
			if (connectForm.Ok()) {
				connect(connectForm.getAddress());
			}
		}
	};

	private void connect(String address) {
		clearResults();
		if (client.Connect(address)) {
			updateConnectedAddress(address);	
		}
		setStatus();
		enableDisableControls();
	}

	Action disconnect = new AbstractAction("Disconnect", createImageIcon("images/Disconnect.png")) {
		public void actionPerformed(ActionEvent event) {
			simulator.Stop();
			cancelExecuteFlag = true;
			updateConnectedAddress("");
			client.Disconnect();
			enableDisableControls();
			clearResults();
		}
	};

	private void executeCommand() {
		executing();
		String command = queryText.getText().trim();
		if (command.length() == 0) return;
		client.Execute(command);
		// determine if we just subscribed  
		if (client.PubSubId().length() > 0 && client.Action().equals("subscribe")) {
			setStatus();
			setJSON();
			// enter event loop
			waitForPubSubEvent();
			return;
		}
		processResponse();
		doneExecuting();
	}

	Action execute = new AbstractAction("Execute", createImageIcon("images/Execute2.png")) {
		public void actionPerformed(ActionEvent event) {
			executeCommand();
		}
	};

	Action cancelExecute = new AbstractAction("Cancel Executing Query", createImageIcon("images/Stop.png")) {
		public void actionPerformed(ActionEvent event) {
			simulator.Stop();
			cancelExecuteFlag = true;
		}
	};

	Action simulate = new AbstractAction("Simulate") {
		public void actionPerformed(ActionEvent event) {
			simulator.Stop();		
			simulator.Address = connectedAddress;
			//
			simulator.Rows = 100;	
			simulator.Columns = 5;
			simulator.TableName = "T" + System.currentTimeMillis();	
			simulator.Start();
			queryText.setText("subscribe * from " + simulator.TableName);
			executeCommand();
			//
		}
	};

	Action about = new AbstractAction("About") {
		public void actionPerformed(ActionEvent event) {
			if (aboutForm == null) aboutForm = new AboutForm(MainForm.this);	
			aboutForm.setLocationRelativeTo(MainForm.this);
			aboutForm.setVisible(true);
		}
	};
	
	// helpers 
	private void defaultTooltips(Action action) {
		action.putValue(Action.SHORT_DESCRIPTION, action.getValue(Action.NAME)); 
	}

	private ImageIcon createImageIcon(String path) {
		java.net.URL url = getClass().getResource(path);
		if (url == null) return null;
		return new ImageIcon(url);
	}

	private void clearResults() {
		dataset.Clear();
		statusText.setText("");	
		jsonText.setText("");	
	}

	private void updateConnectedAddress(String address) {
        setTitle("PubSubSQL Interactive Query " + address);
		connectedAddress = address;
	}

	private void setStatus() {
		if (client.Ok()) {
			statusText.setForeground(Color.black);
			statusText.setText("ok");
			return;
		}
		statusText.setForeground(Color.red);
		statusText.setText("error\n" + client.Error());
		enableDisableControls();
	}

	private void setJSON() {
		jsonText.setText(client.JSON());						
	}

	private void enableDisableControls() {
		boolean connected = client.Connected();
		connectLocalButton.setEnabled(!connected);
		connectLocalMenu.setEnabled(!connected);
		connectButton.setEnabled(!connected);
		connectMenu.setEnabled(!connected);
		disconnectButton.setEnabled(connected);
		disconnectMenu.setEnabled(connected);
		executeButton.setEnabled(connected);
		executeMenu.setEnabled(connected);
		cancelButton.setEnabled(false); cancelMenu.setEnabled(false);
		simulateMenu.setEnabled(executeMenu.isEnabled());
	}

	private void executing() {
		clearResults();
		cancelExecuteFlag = false;
		queryText.setEnabled(false);
		executeButton.setEnabled(false);
		executeMenu.setEnabled(false);
		cancelButton.setEnabled(true);
		cancelMenu.setEnabled(true);
	}

	private void doneExecuting() {
		queryText.setEnabled(true);
		enableDisableControls();	
	}

	private void processResponse() {
		// check if it is result set
		if (client.RowCount() > 0 && client.ColumnCount() > 0) {
			updateDataset(); 
			//resultsTabContainer.SelectedTab = resultsTab;
		}
		//            
		if (client.Failed());//resultsTabContainer.SelectedTab = statusTab;
		setStatus();
		setJSON();			
	}

	private void waitForPubSubEvent() {
		if (cancelExecuteFlag) { 
			doneExecuting();		
			// just reconnect to avoid unsubscribing
			if (connectedAddress.length() > 0) {
				connect(connectedAddress);
			}
			clearResults();
			return;
		}
		if (client.WaitForPubSub(PUBSUB_TIMEOUT)) updateDataset();
		if (client.Failed()) {
			doneExecuting();		
			return;
		}
		// release control to gui thread and post back to continue polling for pubsub events
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				waitForPubSubEvent();
			}
		});	
	}

	private void updateDataset() {
		if (!(client.RowCount() > 0 && client.ColumnCount() > 0)) return;
		dataset.SyncColumns(client);
		while (client.NextRow() && !cancelExecuteFlag) {
			dataset.ProcessRow(client);
		}	
	}
}
